lib/plugin.js

```diff
@@ -130,72 +128,80 @@ const bind = function(name, pluginBind) {
           return deps.push(el.promise);
         });
       });
-      waitFor = Promise.all(deps);
+      await Promise.all(deps);
     }
-    return waitFor
-      .then(function() {
-        $item.removeClass('emit');
-        let bindPromise = pluginBind($item, item);
-        if (!bindPromise || (typeof(bindPromise.then) === 'function')) {
-          bindPromise = Promise.resolve(bindPromise);
-        }
-        // This is where the "bind results" promise for the current item is stored
-        return $item[0].promise = bindPromise;
-      }).then(function() {
-        // If the plugin has the needed callback, subscribe to server side events
-        // for the current page
-        if (window.plugins[name].processServerEvent) {
-          console.log('listening for server events', $item, item);
-          return forward.init($item, item, window.plugins[name].processServerEvent);
-        }
-      }).catch(e => console.log('plugin emit: unexpected error', e));
-  };
+
+    try {
+
+      $item.removeClass('emit');
+      let bindPromise = $item[0].promise = (async function () {
+        return pluginBind($item, item);
+      })()
+
+      await bindPromise
+
+      // If the plugin has the needed callback, subscribe to server side events
+      // for the current page
+      if (window.plugins[name].processServerEvent) {
+        console.log('listening for server events', $item, item);
+        forward.init($item, item, window.plugins[name].processServerEvent);
+      }
+    } catch (e) {
+      console.log('plugin emit: unexpected error', e)
+    }
+  }
   return fn;
 };
 
-plugin.wrap = function(name, p) {
+plugin.wrap = wrap
+function wrap (name, p) {
   p.emit = emit(p.emit);
   p.bind = bind(name, p.bind);
   return p;
 };
 
-plugin.get = (plugin.getPlugin = function(name, callback) {
+plugin.get = (plugin.getPlugin = async function(name, callback) {
+  if (callback == null) { callback = function() {}; }
   if (window.pluginSuccessor[name]) {
     wiki.log('plugin successor', name, window.pluginSuccessor[name]);
     name = window.pluginSuccessor[name]; 
   }
-  if (loadingScripts[name]) { return loadingScripts[name].then(callback); }
-  loadingScripts[name] = new Promise(function(resolve, _reject) {
-    if (window.plugins[name]) { return resolve(window.plugins[name]); }
-    return getScript(`/plugins/${name}/${name}.js`, function() {
-      let p = window.plugins[name];
+  if (window.plugins[name]) {
+    callback(window.plugins[name])
+    return window.plugins[name]
+  }
+  if (!loadingScripts[name]) { 
+    loadingScripts[name] = (async function() {
+      if (window.plugins[name]) { return window.plugins[name]; }
+      await getScript(`/plugins/${name}/${name}.js`)
+      if (!window.plugins[name]) {
+        await getScript(`/plugins/${name}.js`)
+      }
```

Comment:
Here's another case where the `async...await` is making this backwards compatibility easier to read.

Tho' it's another place that makes the `window.plugins[name]` part conspicuous. In this context it looks like we expect the plugin author to assign to the global `plugins` and we're checking for that side-effect. (Feeling more convinced this comment is unrelated to this PR, but seems worth thinking about while we're here).

Reply:
Yeah, that is exactly what has been happening, this change just makes it easier to understand :sweat_smile: I don't think there is anything we can do to improve it here yet, but this change combined w/ import to load plugins would make it very easy to add an alternate way to register plugins using es6 module exports.